### PR TITLE
Add mypy to lint and add dependencies.py

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #880
* __->__ #877

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
it suuure would be nice to start using lintrunner cough cough (@suo)

File: Makefile [hunk 1]
```diff
@@ -36,6 +37,7 @@ lint:
 	black --check --diff $(PY_FILES)
 	isort --check --diff $(PY_FILES)
 	flake8 $(PY_FILES)
+	mypy --show-error-codes $(MYPY_FILES)
```
Comment (voznesenskym) [inline Makefile:40]: Does this pass CI? I could imagine this causes it to fail for a while. Maybe to keep lint passing, add another make target?

Comment (Chillee) [reply Makefile:40]: I only added `dependencies.py` for now.

File: torchinductor/dependencies.py [hunk 2]
```diff
@@ -14,23 +20,25 @@
 
 log = logging.getLogger(__name__)
 
+Dep = Union["MemoryDep", "StarDep"]
```
Comment (Chillee) [inline torchinductor/dependencies.py:23]: What do folks think?

MemoryDep and StarDep seem to be meaningfully different (i.e. StarDep does not have the notion of an index or a size), and there's a bunch of places that currently are annotated as "MemoryDep" that I suspect can actually hold StarDep too.

Comment (desertfire) [reply torchinductor/dependencies.py:23]: It makes sense, but I would prefer to explicitly write `Dep` as a base class.

Comment (jansel) [reply torchinductor/dependencies.py:23]: +1 to add a base class

**Review by jansel — APPROVED** [review]

**Comment by Chillee** [comment]
cc: @ezyang 🤔 

image

**Pull request merged**